[clang] Improve nested name specifier AST representation

mizvekov opened the pull request:
This is a major change on how we represent nested name qualifications in the AST.

* The nested name specifier itself and how it's stored is changed. The prefixes for types are handled within the type hierarchy, which makes canonicalization for them super cheap, no memory allocation required. Also translating a type into nested name specifier form becomes a no-op. An identifier is stored as a DependentNameType. The nested name specifier gains a lightweight handle class, to be used instead of passing around pointers, which is similar to what is implemented for TemplateName. There is still one free bit available, and this handle can be used within a PointerUnion and PointerIntPair, which should keep bit-packing aficionados happy.
* The ElaboratedType node is removed, all type nodes in which it could previously apply to can now store the elaborated keyword and name qualifier, tail allocating when present.
* TagTypes can now point to the exact declaration found when producing these, as opposed to the previous situation of there only existing one TagType per entity. This increases the amount of type sugar retained, and can have several applications, for example in tracking module ownership, and other tools which care about source file origins, such as IWYU. These TagTypes are lazily allocated, in order to limit the increase in AST size.

This patch offers a great performance benefit.

It greatly improves compilation time for [stdexec](https://github.com/NVIDIA/stdexec). For one datapoint, for `test_on2.cpp` in that project, which is the slowest compiling test, this patch improves `-c` compilation time by about 7.2%, with the `-fsyntax-only` improvement being at ~12%.

This has great results on compile-time-tracker as well:
![image](https://github.com/user-attachments/assets/700dce98-2cab-4aa8-97d1-b038c0bee831)

This patch also further enables other optimziations in the future, and will reduce the performance impact of template specialization resugaring when that lands.

It has some other miscelaneous drive-by fixes.

About the review: Yes the patch is huge, sorry about that. Part of the reason is that I started by the nested name specifier part, before the ElaboratedType part, but that had a huge performance downside, as ElaboratedType is a big performance hog. I didn't have the steam to go back and change the patch after the fact.

There is also a lot of internal API changes, and it made sense to remove ElaboratedType in one go, versus removing it from one type at a time, as that would present much more churn to the users. Also, the nested name specifier having a different API avoids missing changes related to how prefixes work now, which could make existing code compile but not work.

How to review: The important changes are all in `clang/include/clang/AST` and `clang/lib/AST`, with also important changes in `clang/lib/Sema/TreeTransform.h`.

The rest and bulk of the changes are mostly consequences of the changes in API.

PS: TagType::getDecl is renamed to `getOriginalDecl` in this patch, just for easier to rebasing. I plan to rename it back after this lands.

Fixes #136624
Fixes https://github.com/llvm/llvm-project/issues/43179
Fixes https://github.com/llvm/llvm-project/issues/68670
Fixes https://github.com/llvm/llvm-project/issues/92757

QuietMisdreavus commented:
ExtractAPI changes look good.

erichkeane reviewed (commented):
Is there ANYTHING we can do to split this up?  This is so huge that github is refusing to load half the changes.

mizvekov committed:
[PATCH 1/7] [clang] Improve nested name specifier AST representation

This is a major change on how we represent nested name qualifications in the AST.

* The nested name specifier itself and how it's stored is changed. The prefixes for types are handled within the type hierarchy, which makes canonicalization for them super cheap, no memory allocation required. Also translating a type into nested name specifier form becomes a no-op. An identifier is stored as a DependentNameType. The nested name specifier gains a lightweight handle class, to be used instead of passing around pointers, which is similar to what is implemented for TemplateName. There is still one free bit available, and this handle can be used within a PointerUnion and PointerIntPair, which should keep bit-packing aficionados happy.
* The ElaboratedType node is removed, all type nodes in which it could previously apply to can now store the elaborated keyword and name qualifier, tail allocating when present.
* TagTypes can now point to the exact declaration found when producing these, as opposed to the previous situation of there only existing one TagType per entity. This increases the amount of type sugar retained, and can have several applications, for example in tracking module ownership, and other tools which care about source file origins, such as IWYU. These TagTypes are lazily allocated, in order to limit the increase in AST size.

This patch offers a great performance benefit.

It greatly improves compilation time for [stdexec](https://github.com/NVIDIA/stdexec). For one datapoint, for `test_on2.cpp` in that project, which is the slowest compiling test, this patch improves `-c` compilation time by about 7.2%, with the `-fsyntax-only` improvement being at ~12%.

This has great results on compile-time-tracker as well:
![image](https://github.com/user-attachments/assets/700dce98-2cab-4aa8-97d1-b038c0bee831)

This patch also further enables other optimziations in the future, and will reduce the performance impact of template specialization resugaring when that lands.

It has some other miscelaneous drive-by fixes.

About the review: Yes the patch is huge, sorry about that. Part of the reason is that I started by the nested name specifier part, before the ElaboratedType part, but that had a huge performance downside, as ElaboratedType is a big performance hog. I didn't have the steam to go back and change the patch after the fact.

There is also a lot of internal API changes, and it made sense to remove ElaboratedType in one go, versus removing it from one type at a time, as that would present much more churn to the users. Also, the nested name specifier having a different API avoids missing changes related to how prefixes work now, which could make existing code compile but not work.

How to review: The important changes are all in `clang/include/clang/AST` and `clang/lib/AST`, with also important changes in `clang/lib/Sema/TreeTransform.h`.

The rest and bulk of the changes are mostly consequences of the changes in API.

PS: TagType::getDecl is renamed to `getOriginalDecl` in this patch, just for easier to rebasing. I plan to rename it back after this lands.

Fixes #136624
Fixes #147000

mizvekov commented:
> Is there ANYTHING we can do to split this up? This is so huge that github is refusing to load half the changes.

I broke the commit down into others, now the first commit only has the AST and TreeTransform changes.

erichkeane commented:
> > Is there ANYTHING we can do to split this up? This is so huge that github is refusing to load half the changes.
> 
> I broke the commit down into others, now the first commit only has the AST and TreeTransform changes.

Can we split them up in reviews/do reviews as a stack if necessary? reviewing each individual 'commit' in the same review is also really difficult.

erichkeane commented on clang/include/clang/AST/ASTContext.h:
```diff
@@ -1749,34 +1748,51 @@ class ASTContext : public RefCountedBase<ASTContext> {
-      return QualType(PrevDecl->TypeForDecl, 0);
-    }
+  QualType getTypeDeclType(const TypeDecl *Decl) const;
+  QualType getTypeDeclType(const TagDecl *) const = delete;
```
Can we do some sort of comments to explain why these are deleted, so that when someone runs into them, they can see what they should be using  instead?

erichkeane commented on clang/include/clang/AST/NestedNameSpecifierBase.h:
```diff
@@ -0,0 +1,585 @@
+
+    /// Microsoft's '__super' specifier, stored as a CXXRecordDecl* of
+    /// the class it appeared in.
+    Super,
```
"While we are here" :)  It would be REALLY nice if instead of `Super` we spelled this as something more obvious what it is.  Something like`MicrosoftSuper`?  `Super` hits too many other language-keyword-type things for me, and it isn't particularly clear when reading other stuff in this patch that it was what this is.

erichkeane commented on clang/lib/AST/ByteCode/Program.cpp:
```diff
@@ -326,7 +326,7 @@ Record *Program::getOrCreateRecord(const RecordDecl *RD) {
       if (!RT)
         return nullptr;
-      const RecordDecl *BD = RT->getDecl();
+      const RecordDecl *BD = RT->getOriginalDecl()->getDefinitionOrSelf();
```
I'm pretty convinced by now, `getOriginalDefinitionOrSelf` would pay HUGE dividends in this patch :)

mizvekov replied:
I have split up all the trivial changes related to 'getOriginalDecl' into its own patch: https://github.com/llvm/llvm-project/pull/149747

erichkeane reviewed (commented):
For obvious reasons, I wasn't able to make even a reasonably thorough review here, but I came across a few things that seemed like good ideas.  All 4 of the patches are STILL huge.

IMO, unless this can be more bite-sized changes, I don't know how to proceed here.

Sirraide commented on clang/include/clang/AST/NestedNameSpecifier.h:
```diff
@@ -6,534 +6,266 @@
-      return (IdentifierInfo *)Specifier;
-
-    return nullptr;
+auto NestedNameSpecifier::getKind() const -> Kind {
```
Is there a reason all of these functions are defined in a header rather than a .cpp file? Does inlining them actually have that big of a performance impact?

mizvekov replied:
It does yes, this was tested against results from `llvm-compile-time-tracker`.

Sirraide reviewed (commented):
Some comments, but I will say that I haven’t nearly reviewed all of this because it’s rather massive (and I’m also not too familiar w/ everything around NNSs).

mizvekov commented on clang/include/clang/AST/NestedNameSpecifierBase.h:
```diff
@@ -0,0 +1,585 @@
+                             FlagOffset = 1, PtrOffset = FlagBits + FlagOffset,
+                             PtrMask = (1u << PtrOffset) - 1u;
+
+  uintptr_t StoredOrFlag;
```
We can't use PointerIntPair because it needs a complete definition for `Decl`, which can't be made available at this point.

mizvekov replied:
I tried the PointerIntPair approach using a custom PointerLikeTypeTrait: https://github.com/llvm/llvm-project/commit/0bdc690d5ee682848f508d0736a02135658cd214

The result has slightly worse performance than the current hand-rolled implementation:
<img width="1465" height="62" alt="image" src="https://github.com/user-attachments/assets/3f5dd3e1-0e78-44a9-a307-e1b5d6d5ebec" />

In this case, I'd prefer keeping the current approach, as it is not much more complicated.

Sirraide commented on clang/include/clang/AST/ASTContext.h:
```diff
@@ -2068,7 +2080,9 @@ class ASTContext : public RefCountedBase<ASTContext> {
   QualType getRawCFConstantStringType() const {
     if (CFConstantStringTypeDecl)
-      return getTypedefType(CFConstantStringTypeDecl);
+      return getTypedefType(ElaboratedTypeKeyword::None,
+                            /*Qualifier=*/std::nullopt,
+                            CFConstantStringTypeDecl);
```
Makes sense

mizvekov commented:
I have further split up this patch series, now the changes related to the NNS representation are moved last in the patch series: https://github.com/llvm/llvm-project/pull/149748

I have also moved some of the trivial changes into yet another patch: https://github.com/llvm/llvm-project/pull/149747

As a result, this and the other patches in the series are smaller.
Let me know if that's enough, or if you would rather I keep splitting further.

Otherwise, I have addressed all pending reviews, and this is ready for another round.

erichkeane commented on clang/include/clang/AST/ASTTypeTraits.h:
```diff
@@ -307,7 +307,7 @@ class DynTypedNode {
   /// For nodes which represent textual entities in the source code,
   /// return their SourceRange.  For all other nodes, return SourceRange().
-  SourceRange getSourceRange() const;
+  SourceRange getSourceRange(bool IncludeQualifier = false) const;
```
I have a mild preference for this being `getSourceRangeIncludingQualifier` (that is, an overload instead of a default arg)?  Unless it results in an `if/else` tree elsewhere.

erichkeane commented on clang/include/clang/AST/NestedNameSpecifier.h:
```diff
@@ -532,4 +532,4 @@ template <> struct DenseMapInfo<clang::NestedNameSpecifierLoc> {
 } // namespace llvm
 
-#endif // LLVM_CLANG_AST_NESTEDNAMESPECIFIER_H
+#endif // LLVM_CLANG_AST_NESTEDNAMESPECIFIERBASE_H
```
This doesn't seem right?

erichkeane reviewed (approved):
I did as thorough of a review as I could.  a small nit, else this is LGTM once the other reviewers who have commented are happy.

mizvekov commented on clang/include/clang/AST/DynamicRecursiveASTVisitor.h:
```diff
@@ -181,14 +181,14 @@ template <bool IsConst> class DynamicRecursiveASTVisitorBase {
   /// \returns false if the visitation was terminated early, true
   /// otherwise (including when the argument is a Null type).
-  virtual bool TraverseType(QualType T);
+  virtual bool TraverseType(QualType T, bool TraverseQualifier = true);
```
I find the parameter easier to debug, it's easier to trace how the value gets propagated.

With a data member / RAII object, it takes longer to figure out where, in a given stack trace, the value was changed.

There is also precedent for this solution here. For example see `clang-tools-extra/clang-tidy/modernize/UseTrailingReturnTypeCheck.cpp` TraverseTypeLoc implementation, where an analogous parameter was implemented (called 'Elaborated' instead).

mizvekov commented:
> I added comment `// FIXME (GH147000): duplicate diagnostics` to `clang/test/SemaCXX/nested-name-spec.cpp` in #147003. Please remove the comment if that issue is fixed by this patch.

This is actually fixed by changes I left out of this patch, I updated the commit description accordingly.

katzdm commented:
Hey @mizvekov - Just wanted to point out a few C++26 changes to nested-name-specifiers (via P2996), in case it helps in any way to future-proof your design.
- A nested-name-specifier can now be either "_splice-specifier_ `::`" or "`template` _splice-specialization-specifier_ `::`", where _splice-specifier_ is "`[:` _constant-expression_ `:]`" and the _constant-expression_ may be dependent (so we may not yet know at parse-time whether it designates a type or a namespace); _splice-specialization-specifier_ is a _splice-specifier_ followed by a template argument list ([expr.prim.id.qual], [basic.splice]).
- A nested-name-specifier of the form "_namespace-name_ `::`" can now be dependent, e.g.,

```cpp
template <std::meta::info R>
  requires is_namespace(R)
int f() {
  namespace Alias = [:R:];
  return Alias::constant;  // 'Alias::' is a dependent namespace-name ([temp.dep.namespace]).
}
```

Hope that neither of these complicate your refactor! Looks like super impressive work; looking forward to seeing it land.

mizvekov commented:
> Hope that neither of these complicate your refactor! Looks like super impressive work; looking forward to seeing it land.

Thanks, that was very a very helpful summary!

I don't think this changes much about the design here, we will get a few new nested name kinds, and these will be less common than all of the existing ones, so we can put them in the slow path, changing `NamespaceAndPrefixStorage` into something which can store the new kinds, without that affecting the existing stuff significantly.

mizvekov committed:
[PATCH 2/7] [clang] improve NestedNameSpecifier: misc small clang changes (#148012)

Other changes

Second patch in the series starting at
https://github.com/llvm/llvm-project/pull/147835

mizvekov commented:
The HLSL Tests bot failure was some GitHub permission issue when publishing the results of the successful run:
```
2025-08-09T07:06:35.4098240Z Request POST /repos/llvm/llvm-project/check-runs failed with 403: Forbidden
2025-08-09T07:06:35.4098670Z 2025-08-09 02:06:35 -0500 - github.GithubRetry -  INFO - Request POST /repos/llvm/llvm-project/check-runs failed with 403: Forbidden
2025-08-09T07:06:35.4099210Z Traceback (most recent call last):
2025-08-09T07:06:35.4100200Z   File "/Users/cbieneman/actions-runner/_work/_actions/EnricoMi/publish-unit-test-result-action/170bf24d20d201b842d7a52403b73ed297e6645b/macos/../python/publish_test_results.py", line 546, in <module>
2025-08-09T07:06:35.4101240Z     main(settings, gha)
2025-08-09T07:06:35.4101840Z   File "/Users/cbieneman/actions-runner/_work/_actions/EnricoMi/publish-unit-test-result-action/170bf24d20d201b842d7a52403b73ed297e6645b/macos/../python/publish_test_results.py", line 269, in main
2025-08-09T07:06:35.4102530Z     Publisher(settings, gh, gha).publish(stats, results.case_results, conclusion)
2025-08-09T07:06:35.4103150Z   File "/Users/cbieneman/actions-runner/_work/_actions/EnricoMi/publish-unit-test-result-action/170bf24d20d201b842d7a52403b73ed297e6645b/python/publish/publisher.py", line 233, in publish
2025-08-09T07:06:35.4112040Z     data = self.publish_check(data)
2025-08-09T07:06:35.4112950Z   File "/Users/cbieneman/actions-runner/_work/_actions/EnricoMi/publish-unit-test-result-action/170bf24d20d201b842d7a52403b73ed297e6645b/python/publish/publisher.py", line 461, in publish_check
2025-08-09T07:06:35.4114640Z     check_run = self._repo.create_check_run(name=self._settings.check_name,
2025-08-09T07:06:35.4115270Z   File "/Users/cbieneman/actions-runner/_work/llvm-project/llvm-project/enricomi-publish-action-venv/lib/python3.9/site-packages/github/Repository.py", line 3885, in create_check_run
2025-08-09T07:06:35.4129000Z     headers, data = self._requester.requestJsonAndCheck(
2025-08-09T07:06:35.4129600Z   File "/Users/cbieneman/actions-runner/_work/llvm-project/llvm-project/enricomi-publish-action-venv/lib/python3.9/site-packages/github/Requester.py", line 550, in requestJsonAndCheck
2025-08-09T07:06:35.4133670Z     return self.__check(*self.requestJson(verb, url, parameters, headers, input, self.__customConnection(url)))
2025-08-09T07:06:35.4134330Z   File "/Users/cbieneman/actions-runner/_work/llvm-project/llvm-project/enricomi-publish-action-venv/lib/python3.9/site-packages/github/Requester.py", line 715, in requestJson
2025-08-09T07:06:35.4134930Z     return self.__requestEncode(cnx, verb, url, parameters, headers, input, encode)
2025-08-09T07:06:35.4135520Z   File "/Users/cbieneman/actions-runner/_work/llvm-project/llvm-project/enricomi-publish-action-venv/lib/python3.9/site-packages/github/Requester.py", line 812, in __requestEncode
2025-08-09T07:06:35.4136240Z     status, responseHeaders, output = self.__requestRaw(cnx, verb, url, requestHeaders, encoded_input)
2025-08-09T07:06:35.4136880Z   File "/Users/cbieneman/actions-runner/_work/llvm-project/llvm-project/enricomi-publish-action-venv/lib/python3.9/site-packages/github/Requester.py", line 846, in __requestRaw
2025-08-09T07:06:35.4137490Z     response = cnx.getresponse()
2025-08-09T07:06:35.4138000Z   File "/Users/cbieneman/actions-runner/_work/llvm-project/llvm-project/enricomi-publish-action-venv/lib/python3.9/site-packages/github/Requester.py", line 198, in getresponse
2025-08-09T07:06:35.4138500Z     r = verb(
2025-08-09T07:06:35.4138940Z   File "/Users/cbieneman/actions-runner/_work/llvm-project/llvm-project/enricomi-publish-action-venv/lib/python3.9/site-packages/requests/sessions.py", line 637, in post
2025-08-09T07:06:35.4141770Z     return self.request("POST", url, data=data, json=json, **kwargs)
2025-08-09T07:06:35.4142390Z   File "/Users/cbieneman/actions-runner/_work/llvm-project/llvm-project/enricomi-publish-action-venv/lib/python3.9/site-packages/requests/sessions.py", line 589, in request
2025-08-09T07:06:35.4142900Z     resp = self.send(prep, **send_kwargs)
2025-08-09T07:06:35.4143390Z   File "/Users/cbieneman/actions-runner/_work/llvm-project/llvm-project/enricomi-publish-action-venv/lib/python3.9/site-packages/requests/sessions.py", line 703, in send
2025-08-09T07:06:35.4143890Z     r = adapter.send(request, **kwargs)
2025-08-09T07:06:35.4144370Z   File "/Users/cbieneman/actions-runner/_work/llvm-project/llvm-project/enricomi-publish-action-venv/lib/python3.9/site-packages/requests/adapters.py", line 667, in send
2025-08-09T07:06:35.4145190Z     resp = conn.urlopen(
2025-08-09T07:06:35.4145680Z   File "/Users/cbieneman/actions-runner/_work/llvm-project/llvm-project/enricomi-publish-action-venv/lib/python3.9/site-packages/urllib3/connectionpool.py", line 934, in urlopen
2025-08-09T07:06:35.4149990Z     retries = retries.increment(method, url, response=response, _pool=self)
2025-08-09T07:06:35.4150560Z   File "/Users/cbieneman/actions-runner/_work/llvm-project/llvm-project/enricomi-publish-action-venv/lib/python3.9/site-packages/github/GithubRetry.py", line 187, in increment
2025-08-09T07:06:35.4152030Z     raise Requester.createException(response.status, response.headers, content)  # type: ignore
2025-08-09T07:06:35.4152720Z github.GithubException.GithubException: 403 {"message": "Resource not accessible by integration", "documentation_url": "https://docs.github.com/rest/checks/runs#create-a-check-run", "status": "403"}
2025-08-09T07:06:35.4812330Z ##[error]Process completed with exit code 1.
```

DavidSpickett commented:
https://github.com/llvm/llvm-project/pull/152993 fixes this.

Pretty sure there's no problem with this PR, it uncovered an existing problem in the formatter itself.

kimgr commented:
Did this PR intentionally break the `RecursiveASTVisitor` API? The new `TraverseQualifier` argument through the whole `Traverse.*Type` hierarchy makes all CRTP overrides not be called, unless I'm missing something.

It's a bit hard to test, because the NNS changes also break a lot of code in IWYU, so it's tricky to play with things in isolation.

mizvekov commented:
> Observation using clangd: In the following construct
> ...
> The "arcana" field of the AST node of "Foo" inside the function reports the "QualType" as "Foo", whereas before it was "N::Foo". Is that expected?

It looks correct in that this is the as-written type.

Now I am not sure clangd wants to present the type that way, maybe they want to fully qualify it for printing instead?
@kadircet @hokein

mysterymath commented:
We are seeing an assertion failure in one of Fuchsia's builders due to this:

```
[38296/176952](67) CXX user.basic_x64-shared/obj/sdk/lib/ld/libld-startup.zircon-startup.cc.o
FAILED: [code=1] user.basic_x64-shared/obj/sdk/lib/ld/libld-startup.zircon-startup.cc.o
../../prebuilt/third_party/clang/custom/bin/clang++ -MD -MF user.basic_x64-shared/obj/sdk/lib/ld/libld-startup.zircon-startup.cc.o.d -o user.basic_x64-shared/obj/sdk/lib/ld/libld-startup.zircon-startup.cc.o -DWITH_FRAME_POINTERS=1 -D_LIBCPP_DISABLE_VISIBILITY_ANNOTATIONS -D_LIBCPP_ENABLE_THREAD_SAFETY_ANNOTATIONS=1 -DZX_ASSERT_LEVEL=2 -D_ALL_SOURCE -DHAVE_LLVM_PROFDATA=0 -I../../zircon/system/p...
clang++: clang/lib/AST/TemplateBase.cpp:610: clang::TemplateArgumentLoc::TemplateArgumentLoc(ASTContext &, const TemplateArgument &, SourceLocation, NestedNameSpecifierLoc, SourceLocation, SourceLocation): Assertion `QualifierLoc.getNestedNameSpecifier() == Argument.getAsTemplateOrTemplatePattern().getQualifier()' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.	Program arguments: ../../prebuilt/third_party/clang/custom/bin/clang++ -MD -MF user.basic_x64-shared/obj/sdk/lib/ld/libld-startup.zircon-startup.cc.o.d -o user.basic_x64-shared/obj/sdk/lib/ld/libld-startup.zircon-startup.cc.o -DWITH_FRAME_POINTERS=1 -D_LIBCPP_DISABLE_VISIBILITY_ANNOTATIONS -D_LIBCPP_ENABLE_THREAD_SAFETY_ANNOTATIONS=1 -DZX_ASSERT_LEVEL=2 -D_ALL_SOURCE -DHAVE_LLVM_PROFDATA=0 -...
1.	<eof> parser at end of file
2.	../../sdk/lib/ld/startup-load.h:183:15: instantiating function definition 'ld::StartupLoadModule<elfldltl::LocalVmarLoader>::LinkModules<trivial_allocator::BasicLeakyAllocator<trivial_allocator::OwningAllocateFunction<trivial_allocator::PageAllocator<trivial_allocator::ZirconVmar>>>, InitialExecAllocator, (lambda at ../../sdk/lib/ld/zircon-startup.cc:210:23) &, zx::vmar &>'
3.	../../sdk/lib/ld/startup-load.h:289:15: instantiating function definition 'ld::StartupLoadModule<elfldltl::LocalVmarLoader>::LoadDeps<trivial_allocator::BasicLeakyAllocator<trivial_allocator::OwningAllocateFunction<trivial_allocator::PageAllocator<trivial_allocator::ZirconVmar>>>, InitialExecAllocator, (lambda at ../../sdk/lib/ld/zircon-startup.cc:210:23) &, zx::vmar &>'
4.	../../sdk/lib/ld/startup-load.h:269:8: instantiating function definition 'ld::StartupLoadModule<elfldltl::LocalVmarLoader>::EnqueueDeps<trivial_allocator::BasicLeakyAllocator<trivial_allocator::OwningAllocateFunction<trivial_allocator::PageAllocator<trivial_allocator::ZirconVmar>>>, zx::vmar &>'
#0 0x000055b727798b48 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (../../prebuilt/third_party/clang/custom/bin/clang+++0x9396b48)
clang++: error: clang frontend command failed with exit code 134 (use -v to see invocation)
Fuchsia clang version 22.0.0git (https://llvm.googlesource.com/llvm-project 10e146a7161065429629a13f99c179a61ffe7721)
Target: x86_64-unknown-fuchsia
Thread model: posix
InstalledDir: ../../prebuilt/third_party/clang/cu...
```

Can you revert or fix forward? I've no idea whether this helps with debugging at all, but this is the source file that produced the assertion failure in clang: https://cs.opensource.google/fuchsia/fuchsia/+/main:sdk/lib/ld/zircon-startup.cc?q=zircon-startup.cc&ss=fuchsia

kadircet commented:
> > Observation using clangd: In the following construct
> > ...
> > The "arcana" field of the AST node of "Foo" inside the function reports the "QualType" as "Foo", whereas before it was "N::Foo". Is that expected?
> 
> It looks correct in that this is the as-written type.
> 
> Now I am not sure clangd wants to present the type that way, maybe they want to fully qualify it for printing instead? @kadircet @hokein

clangd doesn't really have a preference here, we just surface `-ast-dump` in a editor friendly format, for debugging purposes.

one can repro the situation with `clang -xc++ -fsyntax-only -Xclang=-ast-dump a.cc`

where a.cc is:
```cpp
namespace N {
struct Foo {};
}  // namespace N
void f() {
  using namespace N;
  Foo foo;
  N::Foo x;
}
```

pre this change output looks like:
```
`-FunctionDecl 0x5621625282a0 <line:4:1, line:8:1> line:4:6 f 'void ()'
  `-CompoundStmt 0x562162528d10 <col:10, line:8:1>
    |-DeclStmt 0x5621625283e8 <line:5:3, col:20>
    | `-UsingDirectiveDecl 0x562162528390 <col:3, col:19> col:19 Namespace 0x562162528018 'N'
    |-DeclStmt 0x562162528b88 <line:6:3, col:10>
    | `-VarDecl 0x562162528440 <col:3, col:7> col:7 foo 'Foo':'N::Foo' callinit
    |   `-CXXConstructExpr 0x562162528b60 <col:7> 'Foo':'N::Foo' 'void () noexcept'
    `-DeclStmt 0x562162528cf8 <line:7:3, col:11>
      `-VarDecl 0x562162528c38 <col:3, col:10> col:10 x 'N::Foo' callinit
        `-CXXConstructExpr 0x562162528cd0 <col:10> 'N::Foo' 'void () noexcept'
```

emphasis on `-VarDecl 0x562162528440 <col:3, col:7> col:7 foo 'Foo':'N::Foo' callinit`

after thins change this is now reported as `VarDecl 0x5652a7615860 <col:3, col:7> col:7 foo 'Foo' callinit`

this was handled by https://github.com/llvm/llvm-project/blob/main/clang/lib/AST/TextNodeDumper.cpp#L909-L926, I am not sure if change is intended (i think we should still be able to detect & desugar)

bolshakov-a commented:
> The new `TraverseQualifier` argument through the whole `Traverse.*Type` hierarchy ...

@mizvekov, anyway, it would be good to know if `TraverseQualifier` is a customization point for users who implement their own `RecursiveASTVisitor` subclasses, or the default argument value should always be `true`? There is an assertion inside `TraverseQualifiedTypeLoc`:
https://github.com/llvm/llvm-project/blob/72b53cde1c80672217da963b0150a17b79a8383c/clang/include/clang/AST/RecursiveASTVisitor.h#L1282
However, I don't see where in the world `QualifiedTypeLoc` is created...

mizvekov commented:
> After this commit, the comment in `clang/include/clang/AST/Decl.h` for `clang::TypeDecl::getTypeForDecl`,
> isn't accurate any more, there's no more `ASTContext::getRecordType`.

Thanks for reporting, I'll fix in a follow up patch.

 > Observation using clangd: In the following construct
> this was handled by https://github.com/llvm/llvm-project/blob/main/clang/lib/AST/TextNodeDumper.cpp#L909-L926, I am not sure if change is intended (i think we should still be able to detect & desugar)

Right, the difference here is that before this patch, if you print a fully desugared tag type (that's what the 'aka'  in the diagnostic does), the type is printed fully qualified. That's because the name qualifiers were carried in the 'ElaboratedType' sugar node, and we used to handle this situation specially: we assumed a bare tag type only occurred because of canonicalization.

With this patch the tag types themselves carry the name qualifier, so just removing the top level sugar doesn't make them lose name qualifier information.

We still fully qualify canonical tag types when printing them though, but that now happens because a canonical tag type is a special value which is not identical to any other tag type for the same declaration.

We may still want to fully qualify types for the 'aka' diagnostic, but I deferred that for a follow up patch in order to limit the amount of test churn.

For clangd, I assume the same thing is happening here, for this  arcana" field, you are fully desugaring the type before printing it.

> @mizvekov, anyway, it would be good to know if `TraverseQualifier` is a customization point for users who implement their own `RecursiveASTVisitor` subclasses, or the default argument value should always be `true`?

The TraverseQualifier flag means, 'should I traverse NestedNameSpecifiers?'

If you are simply customizing one of the Traverse* methods which take it, but still delegating to the base class Traverse method, you should just forward that flag along to the base method.

If you are implementing your own Traverse method for a type from scratch, then if this flag is false, you should not traverse the nested name specifier for that type.

> There is an assertion inside `TraverseQualifiedTypeLoc`:
> https://github.com/llvm/llvm-project/blob/72b53cde1c80672217da963b0150a17b79a8383c/clang/include/clang/AST/RecursiveASTVisitor.h#L1282
> 
> 
> However, I don't see where in the world `QualifiedTypeLoc` is created...

QualifiedTypeLoc is created implicitly for QualTypes which have top level qualifiers, and they represent the source locations for type qualifiers (not name qualifiers, different thing.)

This assert here means that a top level type qualifier can't appear within a nested name specifier (but it could appear for example in the underlying type for a typedef).

The C++ grammar doesn't allow you to write const / volatile in a way which applies to an entity in the name qualifier itself.

While such a NestedNameSpecifier could still be produced as part of template argument deduction, we simply remove the top level qualifiers in that case, as the NestedNameSpecifier AST node itself still can't represent it.

mysterymath commented:
> > We are seeing an assertion failure in one of Fuchsia's builders due to this:
> 
> Can you please attach the crash reproducer which clang should have generated? Thanks.

https://storage.googleapis.com/fuchsia-artifacts/builds/8707018386865583409/build-debug/clang-crashreports/zircon-startup-bad996.tar

mizvekov commented:
> @mizvekov, thanks for the clarifying! In fact, it means that at least `TraverseTypeLoc` should not, in general, be given `false` as `TraverseQualifier` argument because (it looks like) nothing inside `RecursiveASTVisitor` changes it to `true` before `TraverseQualifiedTypeLoc` call. Note that it may be called not only from the NNS traversal.

Yeah, it may only given false in the implementation of TraverseNestedNameSpecifier, which for out-of-tree users is only relevant if you are overriding that. If you mistakenly passed true instead, the name qualifiers would be visited twice.

emaxx-google commented:
> > > We are seeing an assertion failure in one of Fuchsia's builders due to this:
> > 
> > 
> > Can you please attach the crash reproducer which clang should have generated? Thanks.
> 
> https://storage.googleapis.com/fuchsia-artifacts/builds/8707018386865583409/build-debug/clang-crashreports/zircon-startup-bad996.tar

Minimized this using C-Vise to 620 bytes: https://pastebin.com/yegBAfpe

mizvekov commented:
> Minimized this using C-Vise to 620 bytes: https://pastebin.com/yegBAfpe

Thanks, that has been fixed here: https://github.com/llvm/llvm-project/pull/153344

bolshakov-a commented:
They should be avoided not always but in certain circumstances, so there should be a new flag field in the visitor class. Not many new code, TBH, but using `TraverseQualifier` was so tempting...

bolshakov-a commented:
The assert doesn't hurt us, as I've said. I just asked if it is a "documented way" of working with the new visitor interface. If it not to be avoided, then we will probably just proceed with the `TraverseQualifier` manipulation.

mizvekov commented:
It would be hard for us to document this as an intended interface, but leave that assert in the way, even if you found yourself a way to avoid it. This wouldn't look very polished.

bolshakov-a commented:
Out of curiosity. The old `getDecl()` methods returned the tag type definition (if present), whereas the new `getOriginalDecl()` returns just some of redeclarations (and not necessarily the first one). What is the reason of such change?

DKLoehr commented:
I don't know if this is the same as the other crash report, but we're also seeing assert failures when building chromium. I've minimized the repro to the following program:
```
template <typename T> struct S {
  using typename T::Ty;
  static Ty Val;
};
template <typename T>
S<T>::Ty S<T>::Val;
```

This fails when running `clang++ -c repro.cc` when clang is built with asserts (`-std=c++20` silences a warning but isn't necessary). The crash looks like this:
```
llvm-project/clang/include/clang/AST/TypeLoc.h:79: T clang::TypeLoc::castAs() const [T = clang::DependentNameTypeLoc]: Assertion `T::isKind(*this)' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.      Program arguments: clang++ -c repro.cc
1.      p.cc:6:19: current parser token ';'
<snip long backtrace>
clang++: error: clang frontend command failed with exit code 134 (use -v to see invocation)
clang version 22.0.0git (https://github.com/llvm/llvm-project.git 91cdd35008e9ab32dffb7e401cdd7313b3461892)
Target: x86_64-unknown-linux-gnu
Thread model: posix
Build config: +assertions
```

mizvekov commented:
> Out of curiosity. The old `getDecl()` methods returned the tag type definition (if present) because they called `getInterestingTagDecl` function internally, whereas the new `getOriginalDecl()` returns just some of redeclarations (and not necessarily the first one). What is the reason of such change?

`getOriginalDecl()` returns exactly the declaration that was found by type lookup when producing the type. If this was a forward declaration, because that was the only thing that was in scope at that point the type was written in source code, that is the declaration that will be returned.

This helps establishing the source file origins of the dependencies of some piece of code.

mizvekov commented:
> Thanks! However, for the following piece of code
> `getOriginalDecl()` called on the pointed-to type gives the forward-declaration from the 3rd line, and the forward-declaration from the 1st line for its canonical type.

Yes that is correct.

The 3rd line is what declaration was found by type lookup at that point, and the canonical declaration has long been established in clang as the first declaration.

mizvekov commented:
> I don't know if this is the same as the other crash report, but we're also seeing assert failures when building chromium. I've minimized the repro to the following program:

Thanks, this is going to be fixed by https://github.com/llvm/llvm-project/pull/153862

vbvictor commented:
Hi!
We are seeing regression in `clang-tidy` in `modernize-type-traits` check https://github.com/llvm/llvm-project/issues/153649. I confirmed locally that on 91cdd35008e9ab32dffb7e401cdd7313b3461892 it is bugged and commit before it is not fc44a4fcd3c54be927c15ddd9211aca1501633e7

bolshakov-a commented on clang/lib/AST/ItaniumMangle.cpp:
```diff
@@ -2586,8 +2561,9 @@ bool CXXNameMangler::mangleUnresolvedTypeOrSimpleId(QualType Ty,
-        cast<InjectedClassNameType>(Ty)->getDecl());
+    mangleSourceNameWithAbiTags(cast<InjectedClassNameType>(Ty)
+                                    ->getOriginalDecl()
+                                    ->getDefinitionOrSelf());
```
Is this `getDefinitionOrSelf()` call necessary here? Shouldn't `InjectedClassNameType` always refer to the definition where the class name is injected to?

mizvekov replied:
Not always.

I mean you could certainly attempt to define an `InjectedClassNameType` whose declaration necessarily is the definition.

This would cause many annoyances, most of them surmountable, as you wouldn't be able to form a type to an arbitrary declaration (which we do offhandedly in error recovery and printing diagnostics), and the type would play by different rules than other tag types.

Though I think at the end, you would come to the problem of demoted definitions, which come up for example when merging definitions from the GMF of different modules. You would have to throw away the ability to represent an InjectedClassNameType pointing to a demoted definition.

bolshakov-a replied:
I've caught an example:
```cpp
template <typename>
struct Tpl;

template <typename>
struct Tpl {
  Tpl(const Tpl&) {}
};
```
The injected name from the copy ctor prototype refers to the forward-declaration from the first two lines. Looks a little strange but Ok.

mizvekov replied:
Are you sure it's not just its canonical type which refers to the first two lines?

bolshakov-a replied:
Heh, interesting. `getAs` function actually returns the canonical type.

mizvekov replied:
getAs only desugars. This could happen with a TemplateSpecializationType which its canonical type is an InjectedClassNameType, since in that case desugar just returns the canonical type, but that is not the case for that constructor, since its using an InjectedClassNameType as written in source code.

bolshakov-a replied:
I've understood: there are specializations of `getAs` for the leaf non-sugar types:

https://github.com/llvm/llvm-project/blob/71925a90c8b713d8fccbfae496e10c68f80b932b/clang/include/clang/AST/Type.h#L3138-L3140

If you call `getAs<InjectedClassNameType>()` on a `Type` which is already `InjectedClassNameType`, you get its canonical type.

bolshakov-a replied:
Btw, it is a drawback of templates that you can easily overlook a specialization.

mizvekov replied:
This will be fixed here: https://github.com/llvm/llvm-project/pull/155028

mizvekov commented:
> We started to see a CI failure around the time this PR landed in https://green.lab.llvm.org/job/llvm.org/job/clang-stage2-Rthinlto/1079/
> Any ideas what might be going on here?

@Xazax-hun this could be the same as https://github.com/llvm/llvm-project/issues/153933

Could you confirm if the hang is limited to assertions enabled llvm builds, and if https://github.com/llvm/llvm-project/pull/153996 is sufficient to fix the problem?

rupprecht commented:
Filed #154270 w/ another crash, which looks different from #153933 / #153996.

shafik commented:
Git bisect also blame this PR for: https://github.com/llvm/llvm-project/issues/154270

mizvekov commented:
> Okay, if this is just a temporary oddity, that's fine. My big concern is that "original" implies "first" and that doesn't match the comment on what's returned. Because temporary measures have a tendency to ossify sometimes, it might make sense to add some more comments to clarify the situation. WDYT?

Yeah sure, it would have been a good idea to add a FIXME explaining this, I will do.

rupprecht commented:
I'm trying to reduce a crash that only reproduces w/ in module builds. Does this sound familiar?

```
assertion failed at clang/lib/AST/RecordLayoutBuilder.cpp:3380 in const ASTRecordLayout &clang::ASTContext::getASTRecordLayout(const RecordDecl *) const: D && "Cannot get layout of forward declarations!"
*** Check failure stack trace: ***
    @     0x557b3889784f  clang::ASTContext::getASTRecordLayout()
    @     0x557b3823cd68  clang::ASTContext::getTypeInfoImpl()
    @     0x557b3823e25c  clang::ASTContext::getTypeInfo()
    @     0x557b3823d855  clang::ASTContext::getPreferredTypeAlign()
    @     0x557b3823bf01  clang::ASTContext::getDeclAlign()
    @     0x557b366c6a06  clang::CodeGen::CodeGenFunction::EmitAutoVarAlloca()
    @     0x557b366c165c  clang::CodeGen::CodeGenFunction::EmitVarDecl()
    @     0x557b366c1120  clang::CodeGen::CodeGenFunction::EmitDecl()
    @     0x557b36760580  clang::CodeGen::CodeGenFunction::EmitDeclStmt()
    @     0x557b36754d50  clang::CodeGen::CodeGenFunction::EmitSimpleStmt()
    @     0x557b3675403e  clang::CodeGen::CodeGenFunction::EmitStmt()
    @     0x557b36761891  clang::CodeGen::CodeGenFunction::EmitCompoundStmtWithoutScope()
    @     0x557b368c47d5  clang::CodeGen::CodeGenFunction::GenerateCode()
    @     0x557b368ef9f6  clang::CodeGen::CodeGenModule::EmitGlobalFunctionDefinition()
    @     0x557b368e6bc8  clang::CodeGen::CodeGenModule::EmitGlobalDefinition()
    @     0x557b368d7786  clang::CodeGen::CodeGenModule::EmitDeferred()
    @     0x557b368d77a2  clang::CodeGen::CodeGenModule::EmitDeferred()
...
    @     0x557b368d77a2  clang::CodeGen::CodeGenModule::EmitDeferred()
    @     0x557b368d77a2  clang::CodeGen::CodeGenModule::EmitDeferred()
    @     0x557b368d41ae  clang::CodeGen::CodeGenModule::Release()
    @     0x557b36a1e14e  (anonymous namespace)::CodeGeneratorImpl::HandleTranslationUnit()
```

michaelrj-google commented:
Hi, I'm helping fix some clang-as-a-library users and I'm running into some issues with the TypePrinter after this commit. Specifically the clang generator in sandboxed-api is broken (https://github.com/google/sandboxed-api/tree/main). After doing the trivial fixes (`getDecl` -> `getOriginalDecl`) some of the tests are failing (https://github.com/google/sandboxed-api/blob/main/sandboxed_api/tools/clang_generator/emitter_test.cc).

There are two specific tests failing in that test, `NamedEnumWithoutTypedef` and `NestedAnonymousStruct`.
`NamedEnumWithoutTypedef` expects
```
"enum Color { kRed, kGreen, kBlue }",
"typedef struct { enum Color member; } B"
```
and is getting
```
"typedef struct { enum Color member; } B"
```

While `NestedAnonymousStruct` expects
```
"struct A { struct { int number; } b; int data; }"
```
and is getting
```
"struct A { struct (unnamed struct at input.cc:2:16) b; int data; }"
```

Related, I think the change to [`void TypePrinter::printUsingBefore(const UsingType *T, raw_ostream &OS)`](https://github.com/llvm/llvm-project/blame/d439c9ea4a19bf783120d8efee5e671316666550/clang/lib/AST/TypePrinter.cpp#L1227) might be causing similar issues. The `getKeywordName` is causing `enum` to get printed when the type is listed for template instantiation, which is causing errors. 

If you need more information feel free to reach out, I'm happy to connect you with the relevant people.

bgra8 commented:
After further testing (at google) we're seeing *lots* of other crashes caused by this revision: https://godbolt.org/z/Y3adj3PY1

The previous revision compiles the reduced successfully. Can you please take a look?

mizvekov commented:
> The previous revision compiles the reduced successfully. Can you please take a look?

Fixed in https://github.com/llvm/llvm-project/pull/155272

aeubanks commented:
fyi this uncovered some mismatching struct/class declarations, which affects mangling in the MSVC C++ ABI, leading to linker undefined symbol errors. unsure why we weren't seeing this before this change

bradking commented on clang/lib/AST/ASTContext.cpp:
```diff
@@ -5254,153 +5253,297 @@ ASTContext::getPredefinedSugarType(PredefinedSugarType::Kind KD) const {
+                           /*Qualifier=*/std::nullopt, TD,
+                           /*OwnsTag=*/false, IsInjected, CanonicalType,
+                           /*WithFoldingSetNode=*/false);
+    TD->TypeForDecl = T;
```
If `::getNonInjectedClassName(TD)` returns `true`, it updates `TD` to point at the non-injected decl.  Then `TD->TypeForDecl = T` mutates the non-injected decl to point at the `TagType` created for the injected decl.  This also leaves the injected decl's `TypeForDecl` unset, so the process repeats every time the injected decl is encountered, and a new `TagType` is allocated each time.

mizvekov replied:
Thanks, indeed that looks like what happens.

mizvekov replied:
This will soon be fixed here: https://github.com/llvm/llvm-project/pull/155347.

mizvekov replied:
Done.

mizvekov commented:
> fyi this uncovered some mismatching struct/class declarations, which affects mangling in the MSVC C++ ABI, leading to linker undefined symbol errors. unsure why we weren't seeing this before this change

@aeubanks 
Could be because this patch fixes the canonicalization for dependent names and dependent template specialization types with regards to the elaborated keyword. That should affect how a nested dependent type is mangled if spelled w/o typename or struct/class keywords. These should be mangled differently because they are not functionally equivalent (ie the kind of lookup performed is different when these are instantiated, which could lead to ODR violations).

mizvekov commented:
@aeubanks can you get us a reproducer?

mizvekov commented:
@michaelrj-google 
> ```
>    ::sapi::v::IntBase<enum A::StorageType> v_type__(type__);
> ```
> 
> which clang errors on.

And that enum type, is that created by Sema, or is that synthesized by code from within your project?

If you are creating that type as a canonical type, or that type is being canonicalized for some reason, then the placement of the enum keyword is controlled by the `SuppressTagKeyword` printing policy.

If you are creating that type yourself within your project, you should use the new TagType ASTContext constructor which takes an ElaboratedKeyword as a parameter. Looks like in that case, you should set it to  `ElaboratedKeyword::None`.

aeubanks commented:
```
$ cat /tmp/a.cc
class A;
struct A;
void f(A*) {}
$ clang-old++ -S -o - /tmp/a.cc -O2 -target x86_64-msvc-windows-pc
...
"?f@@YAXPEAVA@@@Z"
...
$ clang-new++ -S -o - /tmp/a.cc -O2 -target x86_64-msvc-windows-pc
...
"?f@@YAXPEAUA@@@Z"
...
```

notice the `V` -> `U`

mizvekov commented:
@aeubanks 

I see, so that's unfortunate MSVC mangles the struct / class difference.
This is a different problem than I had in mind.

So MSVC mangles according to the tag kind of the first declaration, clang after this changes now mangles it according to the    declaration found by lookup at the point of use.

Sounds like if we want strict MSVC bug-for-bug conformance, we should change that to use the first declaration then.

mizvekov commented:
> (I think this was covered by tests even, right?) 
No, this patch didn't change any existing mangling tests at all, the behavior in question here was untested.

bolshakov-a commented:
> We don't want bug-for-bug conformance, but we do want to be ABI compatible.

It really looks like an MSVC bug. If its mangling depends on whether the first (forward-)declaration of a class in a translation unit uses `class` or `struct` class-key, the TUs cannot be linked together in general.

mizvekov commented:
The MSVC mangling issue is resolved here: https://github.com/llvm/llvm-project/pull/155662

rjmccall commented:
I mean, it's an MSVC bug in the sense that it's non-conformant behavior, but it's non-conformant behavior that has long since become a cornerstone of the MSVC ABI, so the bugginess is no longer relevant. We are obliged to maintain compatibility with it if we want to say we implement the ABI correctly.

nico commented:
FWIW, clang has a warning like that too, it's `-Wmismatched-tags` (part of `-Wmost`, which is part of `-Wall`).
